Controller:change bss_infos to use sMacAddress as key

CoralMalachi opened the pull request:
Currently, test 5.4.1 24G keeps failing.

The problem is that the controller sends unnecessary teardown request to
the agent which causing to tear down all BSSes on that radio.

The reason for this is that bss_info is being stored with a string key in the database. 
The UCC sends the MAC address with uppercase hex digits, but the
network_utils::mac_to_string method uses lowercase.

As a result, the value of bss_info_confs is empty, which causes the num_bsss's value 
be zero and that leads to tearing down action : 

`

    if (num_bsss == 0) {
        if (!autoconfig_wsc_add_m2(tlvwscM1, nullptr)) {
            LOG(ERROR) << "Failed setting M2 attributes";
            return false;
        }
    }
`

The solution is to use sMacAddress as the database key instead of a string.

Fixes #334 
Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
controller:change bss_infos to use sMacAddress as key

Currently, test 5.4.1 24G keeps failing.

The problem is that the controller sends unnecessary teardown request to
the agent which s causing to tear down all BSSes on that radio.

The reason for this is that bss_info is being stored with a string key
in the database. The UCC sends the MAC address with uppercase hex
digits, but the network_utils::mac_to_string method uses lowercase.

As a result, no bss_info_confs is found in the database for that MAC
address, so a teardown is sent.

The solution is to use sMacAddress as the database key instead of a string.

Since the bss_info database uses an unordered_map, we need a hash
function for the key. Add a standard hash function for sMacAddr to
net_struct.h.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout reviewed (commented):
Good that you pushed this :-)

I think you're making this way too complicated. Isn't it sufficient to define `operator ==` in struct sMacAdress?

arnout commented:
> I think you're making this way too complicated. Isn't it sufficient to define `operator ==` in struct sMacAdress?

My bad, that operator exists already. What we need is a specialisation `template <> struct std::hash<sMacAddress>` which is exactly what you do here. Only, it should be defined together with the comparison operator (or better yet, in sMacAddr.h) instead of being passed explicitly to the unordered_map.

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -717,9 +717,61 @@ class db {
+        {
+            std::string s1 = beerocks::net::network_utils::mac_to_string(m.oct);
+            s1.erase(std::remove(s1.begin(), s1.end(), ':'), s1.end());
+            size_t h1 = std::hash<std::string>()(s1);
```
I would create the hash based on a uint64_t:

```cpp
    uint64_t value_to_hash = 0;
    for (size_t byte = 0; byte < sizeof(sMacAddr.oct); byte++) {
        value_to_hash <<= 8;
        value_to_hash += sMacAddr.oct[byte];
    }
    return std::hash(value_to_hash);
}
```

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -760,8 +760,9 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(Socket *sd,
-    const auto &bss_info_confs = database.get_bss_info_configuration(al_mac);
-    uint8_t num_bsss           = 0;
+    const auto &bss_info_confs =
+        database.get_bss_info_configuration(network_utils::mac_from_string(al_mac));
```
Use `tlvwscM1->mac_attr().data` directly instead of converting back from string.

CoralMalachi replied:
yes, I have noticed it later but didn't push yet, I'll upload soon :)

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -717,9 +717,21 @@ class db {
 
     std::shared_ptr<vaps_list_t> m_vap_list;
 
+    struct hash_fn {
```
This definition should move to sMacAddr.h (and the name changed of course). (Note: also the operator== and operator!= definitions should move from net_struct.h to sMacAddr.h - in a separate commit of course).

Also, I think it could be defined as the standard hash function for sMacAddr.

```
template <> struct std::hash<sMacAddress> {
    size_t operator()(const sMacAddr &m) const
    ...
};
```

You might need to play around with the namespaces to get this right.

CoralMalachi replied:
I actually thought like you but sMacAddr.h is an auto-generated file, when I do dist-clean all my changes there will be gone, so how to do this right? edit the YAML file? @arnout

arnout replied:
Ah, you're right, I thought sMacAddr.h was just copied from the src directory.

Indeed, then it is not possible to move it to sMacAddr.h. So instead it should move to net_struct.h.

And the rename to `std::hash` should also be possible of course.

CoralMalachi replied:
I planed to move it to net_struct.h but I see now that you already merged this PR.
if I want to add this do you want me to open another PR or revert this one? @arnout

CoralMalachi replied:
Nevermind I can see you have done this already, thanks 👍

arnout reviewed (approved):
Looks good to me as it is - I'd prefer to move the hash function to the definition of sMacAddr but that's more a nitpick.

This solves the issue in certification, right?

Is there any reason to keep this in draft?

arnout committed:
tools: test_flows.sh: add test for uppercase bss_info

The previous commit fixed an issue with an uppercase MAC address in
bss_info. To avoid that this issue returns, change one of the tests to
use an uppercase MAC address (while others still use lowercase).

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

CoralMalachi commented:
> 
> 
> Looks good to me as it is - I'd prefer to move the hash function to the definition of sMacAddr but that's more a nitpick.
> 
> This solves the issue in certification, right?
> 
> Is there any reason to keep this in draft?

Yes, it solves the issue :) (checked with UCC).
